kaggle_environments/envs/werewolf/werewolf.js

```diff
@@ -0,0 +1,4500 @@
+        'SeerInspectActionDataEntry',
+        'DayExileElectedDataEntry',
+        'WerewolfNightEliminationDataEntry',
+        'SeerInspectResultDataEntry',
+        'DoctorSaveDataEntry',
+        'GameEndResultsDataEntry',
+        'PhaseDividerDataEntry',
+        'DiscussionOrderDataEntry'
+    ]);
+
+    let allEventsIndex = 0;
+    let currentDisplayStep = 0;
+    const processedPhaseEvents = new Set(); // This will track events within a single phase.
+    (environment.info?.MODERATOR_OBSERVATION || []).forEach((stepEvents, kaggleStep) => {
+        (stepEvents || []).flat().forEach(dataEntry => {
+            const event = JSON.parse(dataEntry.json_str);
+            const dataType = dataEntry.data_type;
+            const visibleInUI = event.visible_in_ui ?? true;
+
+            console.log(`[RAW SEEN] Kaggle Step: ${kaggleStep}`, { dataType: dataType, event: event });
+
+            if (!visibleInUI) {
+              return;
+            }
+
+            // 1. Reset our memory on key phase-changing events.
+            if (event.event_name === 'day_start' || event.event_name === 'night_start' || event.description?.includes('Voting phase begins')) {
+                processedPhaseEvents.clear();
+            }
+
+            // 2. Generate a unique "fingerprint" based on essential event content.
+            let eventFingerprint = event.description;
+            const eventData = event.data;
+
+            // 3. Check against the new fingerprint.
+            if (processedPhaseEvents.has(eventFingerprint)) {
+                return; // This is a repeated state update, not a new event. Ignore it.
+            }
+            processedPhaseEvents.add(eventFingerprint);
+
+            // The rest of your original processing logic follows...
+            const isVisibleDataType = visibleEventDataTypes.has(dataType);
+            const isVisibleEntryType = systemEntryTypeSet.has(event.event_name);
+
+            if (!isVisibleDataType && !isVisibleEntryType) {
+                return;
+            }
+
+            // (Your existing code to push events to player.allEvents, player.displayEvents, etc.)
+            event.kaggleStep = kaggleStep;
+            event.dataType = dataType; // Ensure dataType is set
+            player.allEvents.push(event);
+            player.eventToKaggleStep.push(kaggleStep);
+
+            if (dataType !== 'PhaseDividerDataEntry') {
+                player.displayEvents.push(event);
+                player.displayStepToAllEventsIndex.push(allEventsIndex);
+                player.allEventsIndexToDisplayStep[allEventsIndex] = currentDisplayStep;
+                currentDisplayStep++;
+            }
+            allEventsIndex++;
+        });
+    });
+
+    console.log(`[FINAL STEP LIST]`, player.displayEvents);
+
+    const newSteps = player.displayEvents.map((event) => {
+        return player.originalSteps[event.kaggleStep];
+    });
+
+    setTimeout(() => {
+        if (window.kaggle) {
```

Comment:
this feels a bit risky - is there a nicer way to make sure this is synced from the player.html file?

Reply:
There are two main issues:
(1) Kaggle step vs game step inconsistency. This I think would be tougher to solve since we already shoehorned the game logic into Kaggle steps/interpreter call. I can work with Kaggle to address this design after launch.
(2) How to properly support control functions overrides? For this I seek guidance from your end.